File: 5. Longest Palindromic Substring/solution.py
def preprocess(s: str) -> str:
    return '#' + '#'.join(s) + '#'

def palindrom_sub(s: str) -> str:
    if len(s) < 1:
        return ""

    processed_s = preprocess(s)
    n = len(processed_s)
    P = [0] * n
    C, R = 0, 0

    for i in range(1, n - 1):
        mirror = 2 * C - i
        
        if i < R:
            P[i] = min(R - i, P[mirror])
        
        while i + (1 + P[i]) < n and i - (1 + P[i]) >= 0 and processed_s[i + (1 + P[i])] == processed_s[i - (1 + P[i])]:
            P[i] += 1
        
        if i + P[i] > R:
            C = i
            R = i + P[i]
    
    max_len = max(P)
    center_index = P.index(max_len)
    
    start_index = (center_index - max_len) // 2
    return s[start_index:start_index + max_len]


print(palindrom_sub("babad"))
print(palindrom_sub("cbbd"))

File: 5. Longest Palindromic Substring/solution.go
package main

import (
	"fmt"
	"math"
	"strings"
)

func main() {
	fmt.Println(palindromSub("babad"))
	fmt.Println(palindromSub("cbbd"))
}

func preprocess(s string) string {
	return "#" + strings.Join(strings.Split(s, ""), "#") + "#"
}

func palindromSub(s string) string {
	if len(s) < 1 {
		return ""
	}

	processedS := preprocess(s)
	n := len(processedS)
	p := make([]int, n)
	c, r := 0, 0

	for i := 1; i < n-1; i++ {
		mirror := 2*c - i

		if i < r {
			p[i] = int(math.Min(float64(r-i), float64(p[mirror])))
		}

		for i+(1+p[i]) < n && i-(1+p[i]) >= 0 && processedS[i+(1+p[i])] == processedS[i-(1+p[i])] {
			p[i]++
		}

		if i+p[i] > r {
			c = i
			r = i + p[i]
		}
	}

	maxLen := 0
	centerIndex := 0

	for i, num := range p {
		if num > maxLen {
			maxLen = num
			centerIndex = i
		}
	}

	startIdx := (centerIndex - maxLen) / 2
	return s[startIdx : startIdx+maxLen]
}
